### PR TITLE
Fix ICE when rustdoc is scraping examples inside of a proc macro

### LINKED ISSUE #90567: Ice with rustdoc scrape_examples
When running 
```
cargo +nightly doc -Z unstable-options -Z rustdoc-scrape-examples=examples
```
on the `ice` branch of https://github.com/mejrs/pyo3/tree/ice , I get the following error:

```text
~/pyo3$ cargo +nightly doc -Z unstable-options -Z rustdoc-scrape-examples=examples
   Compiling pyo3 v0.14.5 (/home/bruno/pyo3)
thread 'rustc' panicked at 'assertion failed: enclosing_item_span.contains(expr_span)', src/librustdoc/scrape_examples.rs:93:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

error: Unrecognized option: 'scrape-examples-output-path'

error: could not document `pyo3`

Caused by:
  process didn't exit successfully: `rustdoc --edition=2018 --crate-type bin --crate-name word_count examples/word-count/src/lib.rs -o /home/bruno/pyo3/target/doc --cfg 'feature="auto-initialize"' --cfg 'feature="default"' --cfg 'feature="extension-module"' --cfg 'feature="indoc"' --cfg 'feature="macros"' --cfg 'feature="paste"' --cfg 'feature="pyo3-macros"' --cfg 'feature="unindent"' --error-format=json --json=diagnostic-rendered-ansi -C metadata=4cd6a3c2c10c880b -Zunstable-options --scrape-examples-output-path /home/bruno/pyo3/target/debug/deps/pyo3-87828202807095e6.examples --scrape-examples-target-crate pyo3-macros --scrape-examples-target-crate pyo3-macros-backend --scrape-examples-target-crate pyo3-build-config --scrape-examples-target-crate pyo3-benchmarks --scrape-examples-target-crate pyo3 --scrape-examples-target-crate pyo3-pytests --scrape-examples-target-crate maturin-starter --scrape-examples-target-crate setuptools-rust-starter --scrape-examples-target-crate word-count -L dependency=/home/bruno/pyo3/target/debug/deps --extern assert_approx_eq=/home/bruno/pyo3/target/debug/deps/libassert_approx_eq-73f29b3f801eac6d.rmeta --extern bitflags=/home/bruno/pyo3/target/debug/deps/libbitflags-c30f35abb0b74f97.rmeta --extern cfg_if=/home/bruno/pyo3/target/debug/deps/libcfg_if-88879d2dc0fa106c.rmeta --extern criterion=/home/bruno/pyo3/target/debug/deps/libcriterion-a1abf6832c975b25.rmeta --extern half=/home/bruno/pyo3/target/debug/deps/libhalf-b6e268e7a94f2f0e.rmeta --extern indoc=/home/bruno/pyo3/target/debug/deps/libindoc-7a3b6554d729eba7.rmeta --extern libc=/home/bruno/pyo3/target/debug/deps/liblibc-56b7c085aca1bd82.rmeta --extern parking_lot=/home/bruno/pyo3/target/debug/deps/libparking_lot-28e0a32b822985ad.rmeta --extern paste=/home/bruno/pyo3/target/debug/deps/libpaste-50cadf247695eb32.rmeta --extern proptest=/home/bruno/pyo3/target/debug/deps/libproptest-011787d4a2170b22.rmeta --extern pyo3=/home/bruno/pyo3/target/debug/deps/libpyo3-fd794029b70a5af7.rmeta --extern pyo3=/home/bruno/pyo3/target/debug/deps/libpyo3-fd794029b70a5af7.rmeta --extern pyo3_macros=/home/bruno/pyo3/target/debug/deps/libpyo3_macros-74b16cc6327df15f.so --extern rayon=/home/bruno/pyo3/target/debug/deps/librayon-6402c0587be315bb.rmeta --extern rustversion=/home/bruno/pyo3/target/debug/deps/librustversion-69c11d077ef44fc2.so --extern serde_json=/home/bruno/pyo3/target/debug/deps/libserde_json-28ebb15e8db233fc.rmeta --extern trybuild=/home/bruno/pyo3/target/debug/deps/libtrybuild-5db6156090fb714a.rmeta --extern unindent=/home/bruno/pyo3/target/debug/deps/libunindent-a67f785cc7bb6fd1.rmeta --crate-version 0.14.5 --cfg Py_3_6 --cfg Py_3_7 --cfg Py_3_8 --cfg 'py_sys_config="WITH_THREAD"' --cfg track_caller --cfg min_const_generics --cfg addr_of` (exit status: 1)

```

(this is quite a big library and I couldn't figure out a minimal example in the time I had)

### PR DESCRIPTION
**Pull request by willcrichton**
This PR provides a clearer semantics for how --scrape-examples interacts with macros. If an expression's span AND it's enclosing item's span both are not `from_expansion`, then the example will be scraped. The added test case `rustdoc-scrape-examples-macros` shows a variety of situations.

* A macro-rules macro that takes a function call as input: good
* A macro-rules macro that generates a function call as output: bad
* A proc-macro that generates a function call as output: bad
* An attribute macro that generates a function call as output: bad
* An attribute macro that takes a function call as input: good, if the proc macro is designed to propagate the input spans

I ran this updated rustdoc on pyo3 and confirmed that it successfully scrapes examples from inside a proc macro, eg

Screen Shot 2021-11-04 at 1 11 28 PM

(cc @mejrs) 

Additionally, this PR fixes an ordering bug in the highlighting logic.

Fixes https://github.com/rust-lang/rust/issues/90567.

r? @jyn514

### CODE REVIEW

**Comment by jyn514** [comment]
> An attribute macro that takes a function call as input: good, if the proc macro is designed to propagate the input spans

Can you also test what happens if it's buggy and doesn't propagate spans? It seems fine to discard that example, I just want to make sure it doesn't cause an ICE.

File: src/librustdoc/html/highlight.rs [hunk 1]
```diff
@@ -270,12 +270,18 @@ struct Decorations {
...
+        // Extract tuples (start, end, kind) into separate sequences of (start, kind) and (end).
+        let (mut starts, mut ends): (Vec<_>, Vec<_>) = info
             .0
             .into_iter()
             .map(|(kind, ranges)| ranges.into_iter().map(move |(lo, hi)| ((lo, kind), hi)))
             .flatten()
             .unzip();
+
+        // Sort the sequences in document order.
+        starts.sort_by_key(|(i, _)| *i);
```
Comment (jyn514) [inline src/librustdoc/html/highlight.rs:282]:
```suggestion
        starts.sort_by_key(|(lo, _)| *lo);
```

**Comment by camelid** [comment]
> Additionally, this PR fixes an ordering bug in the highlighting logic.

Could you separate that into another commit? It'll make easier to understand why code was changed later.

And thanks for adding a test ❤️

File: src/librustdoc/scrape_examples.rs [hunk 2]
```diff
@@ -168,13 +162,30 @@ where
...
+            "Attempted to scrape call at [{:?}] whose enclosing item [{:?}] doesn't contain the span of the call.",
+            span,
+            enclosing_item_span
+        );
+
         // Save call site if the function resolves to a concrete definition
         if let ty::FnDef(def_id, _) = ty.kind() {
             // Ignore functions not from the crate being documented
             if self.target_crates.iter().all(|krate| *krate != def_id.krate) {
+                trace!("Rejecting expr not from crate whitelist: {:?}", span);
```
Comment (jyn514) [inline src/librustdoc/scrape_examples.rs:188]:
```suggestion
                trace!("Rejecting expr from crate not being documented: {:?}", span);
```
and maybe remove the comment above, it seems redundant with the logging.

File: src/librustdoc/scrape_examples.rs [hunk 3]
```diff
@@ -240,6 +252,13 @@ crate fn run(
...
         tcx.hir().visit_all_item_likes(&mut finder.as_deep_visitor());
 
+        // Sort call locations within a given file in document order
+        for fn_calls in calls.values_mut() {
+            for file_calls in fn_calls.values_mut() {
+                file_calls.locations.sort_by_key(|loc| loc.call_expr.byte_span.0);
+            }
+        }
+
         // Save output to provided path
```
Comment (jyn514) [inline src/librustdoc/scrape_examples.rs:262]: Are you sure the order is preserved when serializing and deserializing? Maybe it makes sense to sort it during deserialization instead?

Please add a test for this either way.

**Comment by jyn514** [comment]
For posterity, this fixes the following ICE when running on Pyo3 (reported on discord):
```
$ cargo +nightly doc -Z unstable-options -Z rustdoc-scrape-examples=examples
   Compiling pyo3 v0.14.5 (C:\Users\user\rust\pyo3)
thread 'rustc' panicked at 'assertion failed: enclosing_item_span.contains(expr_span)', src\librustdoc\scrape_examples.rs:93:9
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

error: Unrecognized option: 'scrape-examples-output-path'

error: could not document `pyo3
```

**Comment by camelid** [comment]
(FYI, I just created an `A-rustdoc-scrape-examples` label to use in tracking this feature. I'll add it to some other PRs and issues too.)

File: src/librustdoc/scrape_examples.rs (hunk 3 above)
Comment (willcrichton) [reply src/librustdoc/scrape_examples.rs:261]: I'm assuming that the serialization library would serialize / deserialize a vector in the order as given (but not, say a hashmap).

I updated the `rustdoc-scrape-examples-ordering` test to check that the `focus` class is applied to the first example in a file.

**Comment by jyn514** [comment]
r=me with CI passing :)

@bors delegate=willcrichton

**Comment by mejrs** [comment]
Is this supposed to draw code examples from `/tests`? That doesn't sound right.

**Comment by jyn514** [comment]
> Is this supposed to draw code examples from `/tests`? That doesn't sound right.

@mejrs I suspect @willcrichton was passing some magic flags to cargo to avoid having to modify Pyo3 - this PR can't affect which examples rustdoc looks at, that's something cargo controls.

**Comment by willcrichton** [comment]
@mejrs pyo3 doesn't have examples configured, i.e. the `--examples` filter to Cargo is empty. So I used the `rustdoc-scrape-examples=all` which scrapes examples from your entire workspace.

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/90592#issuecomment-961547016
@bors r-

**Comment by jyn514** [comment]
>   = note: arm-none-eabi-gcc: error: unrecognized command line option '-m64'

Ugh, this again. @willcrichton try moving the test to run-make-fulldeps. You can still use the included makefile with `include ../run-make/scrape.mk`.

**Comment by willcrichton** [comment]
Ok done. This test doesn't actually use `scrape.mk` so no changes needed besides moving the test.

**Comment by jyn514** [comment]
@willcrichton every platform names the suffix for dynamic libraries something different - you can either try to make it work in the makefile (with `$(RUSTC) --crate-name foo --crate-type dylib --print file-names`, see the trouble I've been having in https://github.com/rust-lang/rust/pull/88368/files#diff-33be197db3e3ddb089ba3380ea3d6ce10944ff5517531547039f0fa6c06216e3) or just make it linux-onluy.

**Comment by willcrichton** [comment]
Ok I did your suggested change @jyn514.

**Comment by jyn514** [comment]
@bors r+

Thanks!

**Comment by camelid** [comment]
I guess fetching the span was quite slow and moving it to after the expansion check (and other checks) improved performance?

**Pull request merged**